ipam/pool_test.go

```diff
@@ -0,0 +1,1007 @@
+						Id:     poolId,
+						IfName: ifName,
+					}
+					options := map[string]string{}
+					options[OptInterfaceName] = "en0"
+					ap, err := as.requestPool("", "", options, false)
+					Expect(err).To(Equal(errNoAvailableAddressPools))
+					Expect(ap).To(BeNil())
+				})
+			})
+
+			Context("When addressSpace has one pool available", func() {
+				It("Should return the pool", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					poolId := "10.0.0.0/16"
+					as.Pools[poolId] = &addressPool{
+						Id: poolId,
+					}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal(poolId))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+
+			Context("When addressSpace has pools with different priority", func() {
+				It("Should return the pool with the highest priority", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					as.Pools["10.0.0.0/16"] = &addressPool{
+						Id:       "10.0.0.0/16",
+						Priority: 1,
+					}
+					as.Pools["10.1.0.0/16"] = &addressPool{
+						Id:       "10.1.0.0/16",
+						Priority: 2,
+					}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal("10.1.0.0/16"))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+
+			Context("When addressSpace has pools with different addresses", func() {
+				It("Should return the pool with the highest number of addresses", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					as.Pools["10.0.0.0/16"] = &addressPool{
+						Id:        "10.0.0.0/16",
+						Addresses: map[string]*addressRecord{},
+					}
+					as.Pools["10.1.0.0/16"] = &addressPool{
+						Id:        "10.1.0.0/16",
+						Addresses: map[string]*addressRecord{},
+					}
+					as.Pools["10.1.0.0/16"].Addresses["10.1.0.1/16"] = &addressRecord{}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal("10.1.0.0/16"))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+		})
+
+		Describe("Test releasePool", func() {
+			Context("When pool not found", func() {
+				It("Should raise an error", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					err := as.releasePool("10.0.0.0/16")
+					Expect(err).To(Equal(errAddressPoolNotFound))
+				})
+			})
+
+			Context("When pool is not in use", func() {
+				It("Should raise an error", func() {
+					poolId := "10.0.0.0/16"
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					as.Pools[poolId] = &addressPool{
+						RefCount: 0,
+					}
+					err := as.releasePool("10.0.0.0/16")
+					Expect(err).To(Equal(errAddressPoolNotInUse))
+					Expect(as.Pools[poolId]).NotTo(BeNil())
+				})
+			})
+
+			Context("When pool's epoch is less than the space's epoch and pool is never in use", func() {
```

Comment:
```suggestion
			Context("When pool's epoch is less than the space's epoch and pool is in use", func() {
```

Reply:
In this case, by `ap.RefCount--`, this pool is not in used and should be deleted.

Reply:
RefCount: 1 so it mean pool in use  by one entity and should be deleted. If pool not in use, delete should fail

Reply:
Yes we first release the address, and reduce the refcount, ant then the refcount will be 0 and not in use, should delete